### DISCUSSION TITLE
A serious geometry issue in DAGMC

### DISCUSSION THREAD

**Discussion by LOUWEE** [svalinn/DAGMC#881] [General]
In our recent experience with DAGMC in MCNP6.2 version, an issue caught us and made us have to redo two week's work. The problem started with a simple overlapping in generating the geometry. The overlapping surfaces were so big, any particle in MCNP transport was supposed to hit the surface and report a loss of particle.. However, in flooding 1 billion particles, no particle was reported. We only caught this at the end of the full calculations, results of which showed some weird areas with zero tallied fluxes though they were not supposed to be zero at all. By using PTRAC, we found the particle passed the overlapping surfaces and disappeared when inside the cell containing the overlapping surface and reappeared when it traveled outside of that cell. I guess that is why it did not report a loss of particle. However, this issue would cause a serious dent in our confidence in using DAGMC. Hope it could be solved fast.
Thanks.

Wei

**Comment by gonuke** [comment]
Thank you for reporting your experience with this issue.  Are you able to share your model so that we can examine why this happened?  Or can you create a simple model that exhibits this behavior that you can share?

It is true that we expect there to be no overlaps in the models that DAGMC uses and we expect the model builders to ensure that there are no overlaps.  However, we are always interested in confirming where users to experience such problems.

**Comment by LOUWEE** [comment]
Thanks. I can certainly share the model and you may let me know where to upload. One thing surprised me that DAGMC itself can not detect overlapping geometry problems?

Wei

**Reply by pshriwise** [reply]
Hi @LOUWEE

We do have an overlap detection program that should be built with DAGMC now -- it should be located in the same directory as other DAGMC tools like `check/make_watertigh` for your DAGMC installation. This tool can be passed an `.h5m` file and report any discovered overlaps. There are improvements I'd like to make to this tool to make it a little more robust (see #797), but for an overlap as large as the one you're describing I'd hope that it would be detected by this tool.

Hope this helps!

-Patrick

**Reply by LOUWEE** [reply]
Thanks Patrick,

Please let us know which version of DAGMC we should download and I am happy to perform a test.

Wei

**Reply by pshriwise** [reply]
Hi Wei,

This tool is available as part of DAGMC since v3.2.0 ([release notes here](https://github.com/svalinn/DAGMC/releases/tag/3.2.0)), so any version 3.2.0 or newer should include that tool.

Patrick

**Reply by LOUWEE** [reply]
Thanks Patrick,

I tested the overlap tool. It did fairly well in detecting overlapping for my problem if it was limited to the component I am aware of. However, if it was applied to the complete model, the speed is really an issue. It takes more than an hour to complete just 1%. I specified 10 points along the edge and 16 threads. Is there a parallel version?

Wei

**Reply by pshriwise** [reply]
Hi Wei,

Is double-down enabled in your DAGMC installation?

Patrick

**Comment by LOUWEE** [comment]
Patrick,

What is double-down? I can ask Kyle about it.

Wei

From: Patrick Shriwise ***@***.***>
Date: Friday, June 23, 2023 at 6:31 AM
To: svalinn/DAGMC ***@***.***>
Cc: Lu, Wei ***@***.***>, Mention ***@***.***>
Subject: [EXTERNAL] Re: [svalinn/DAGMC] A serious geometry issue in DAGMC (Discussion #881)

Hi Wei,

Is double-down enabled in your DAGMC installation?

Patrick

**Reply by pshriwise** [reply]
Double-down is an interface that allows DAGMC to use Intel's Embree raytracing toolkit. It gives a pretty big boost to performance in both model initialization and ray tracing performance. It tends to get much better scaling than using MOAB's native ray tracer (what DAGMC relies on by default).

https://github.com/pshriwise/double-down

Threaded particle tracking study with OpenMC:

The "MOAB" entry here is DAGMC using the default ray tracer in MOAB; the "MOAB w/ Embree" is DAGMC built with double-down.

![atr_scaling](https://github.com/svalinn/DAGMC/assets/4563941/c0a17b69-5951-439a-8a8b-98a28d93cbd2)
